drivers: display: ili9xxx: Add support for ili9342c display controller

ExtremeGTX opened the pull request:
Driver for ili9342c display controller 
Tested on [m5stack Core Basic](https://shop.m5stack.com/products/basic-core-iot-development-kit?variant=16804801937498) using `sample/subsys/display/lvgl`

ExtremeGTX commented on drivers/display/display_ili9xxx.c:
```diff
@@ -314,6 +317,11 @@ static int ili9xxx_configure(const struct device *dev)
 		return r;
 	}
 
+	r = config->regs_init_fn(dev);
+	if (r < 0) {
+		return r;
+	}
+
```
I moved `regs_init_fn` call here because applying display configuration before init doesn't result in the desired effect.
Would be great if someone has one of the other display controllers `9341` or `9488` to verify that this change didn't affect them.

ExtremeGTX replied:
it looks like something was up with my Env, i reverted this change.

diegoherranz commented:
I was about to submit a nearly identical pull request. I'll try to review yours and provide comments.
Thanks!

diegoherranz commented on dts/bindings/display/ilitek,ili9342c.yaml:
```diff
@@ -0,0 +1,111 @@
+# Copyright (c) 2018, Jan Van Winkel <jan.van_winkel@dxplore.eu>
```
I've commented about not using the default values on a few registers in this file. In order to avoid making too many comments, the same thing applies to any register in this file. It should follow the default in the datasheet unless there is a good reason to deviate.

diegoherranz commented on drivers/display/display_ili9342c.c:
```diff
@@ -0,0 +1,94 @@
+	if (r < 0) {
+		return r;
+	}
+
```
etmod missing here (defined in ili9342c.h)

diegoherranz commented on drivers/display/display_ili9342c.c:
```diff
@@ -0,0 +1,94 @@
+		return r;
+	}
+
+	LOG_HEXDUMP_DBG(regs->pwctrl3, ILI9342C_PWCTRL3_LEN, "PWCTRL3");
```
Maybe place next to pwrctrl1 and 2?

diegoherranz commented on drivers/display/display_ili9342c.c:
```diff
@@ -0,0 +1,94 @@
+/*
```
Unless there is a special reason (for instance SETEXTC going first), it would be good to follow the register address order.

diegoherranz commented on drivers/display/display_ili9342c.c:
```diff
@@ -0,0 +1,94 @@
+	const struct ili9xxx_config *config = dev->config;
+	const struct ili9342c_regs *regs = config->regs;
+	int r;
+
```
GAMSET missing here. Unless ili9xxx does it?

diegoherranz reviewed (changes requested):
I've had a look and provided some comments. Let me know if you have any questions.
Thanks!

ExtremeGTX commented:
@diegoherranz Thank you for the review comments 👍, I have fixed all of them, and also retested orientation modes.
I don't like the use of `#ifdef` in there, but I don't have the other ili9xxx chips to play around with these values.

diegoherranz commented:
> @diegoherranz Thank you for the review comments 👍, I have fixed all of them, and also retested orientation modes. I don't like the use of `#ifdef` in there, but I don't have the other ili9xxx chips to play around with these values.

Thanks for making these changes. I haven't forgotten about this one, simply I haven't had time to re-review it lately. I will as soon as possible.

jfischer-no commented on drivers/display/display_ili9xxx.c:
```diff
@@ -250,14 +250,24 @@ static int ili9xxx_set_orientation(const struct device *dev,
 	uint8_t tx_data = ILI9XXX_MADCTL_BGR;
 
 	if (orientation == DISPLAY_ORIENTATION_NORMAL) {
+#if !CONFIG_ILI9342C
```
Commit message does not explain why it is necessary. Anyway, right way could be
```c

	if (orientation == DISPLAY_ORIENTATION_NORMAL &&
	    !IS_ENABLED(CONFIG_ILI9342C)) {

```

But it breaks support of multiple instances. Maybe specific compatibles can be a solution, see https://github.com/zephyrproject-rtos/zephyr/pull/48163

ExtremeGTX replied:
I didn't get the solution exactly, i can see for ili9xxx series there are specific compatibles per controller model like:
`ilitek,ili9340`, `ilitek,ili9341`, `ilitek,ili9488`.
The problem is this model 9342c has different Orientation settings than others, as an example [See here](https://github.com/letscontrolit/ESPEasy/blob/5f40dafc5ac315ea4ddd623531d71de33218506d/lib/Adafruit_ILI9341/Adafruit_ILI9341.cpp#L623)

What do you think if an orientation function implemented per model as part of [config struct](https://github.com/zephyrproject-rtos/zephyr/blob/fb60aab2458684f347533cba3658c89757580932/drivers/display/display_ili9xxx.h#L57) ? then that function will be called by `ili9xxx_set_orientation` ?

jfischer-no replied:
The first step should be to identify the reason for the differences, MADCTL(0x36) register description looks the same for example for ILI9340 and ILI9342C. You can then use DT_HAS_COMPAT_STATUS_OKAY(ilitek,ili9342c) to add a quirk to specific driver instance or just set a boolean in the driver's config.

ExtremeGTX replied:
I tried to identify why this controller behave differently but couldn't find a clear answer. 
According to the datasheet it should be the same MADCTL command used in the other models, but dunno why it behaves differently, checked other drivers on Github also, all they do is handling this controller differently.
Also, in the datasheet the part about init code is not in 9342c datasheet.
So I think the solution would be to add a quirk to the driver like the example you mentioned above.

ExtremeGTX replied:
The mystery of different behavior of MADCTL in this specific model not solved yet, I even asked m5stack and they don't have an answer.
So, what is the Zephyr way here to tell the user orientation is not support message?

jfischer-no commented on drivers/display/display_ili9342c.h:
```diff
@@ -0,0 +1,121 @@
+
+/* Initializer macro for ILI9342C registers. */
+#define ILI9342c_REGS_INIT(n)									\
+	BUILD_ASSERT(DT_PROP_LEN(DT_INST(n, ilitek_ili9342c), gamset) == ILI9342C_GAMSET_LEN, \
+		     "ILI9342c: Error length of gamma set (GAMSET) register");                    \
+	BUILD_ASSERT(DT_PROP_LEN(DT_INST(n, ilitek_ili9342c), ifmode) == ILI9342C_IFMODE_LEN, \
+		     "ILI9342c: Error length of frame rate control (IFMODE) register");           \
+	BUILD_ASSERT(DT_PROP_LEN(DT_INST(n, ilitek_ili9342c), frmctr1) == ILI9342C_FRMCTR1_LEN, \
+		     "ILI9342c: Error length of frame rate control (FRMCTR1) register");          \
+	BUILD_ASSERT(DT_PROP_LEN(DT_INST(n, ilitek_ili9342c), invtr) == ILI9342C_INVTR_LEN,   \
+		     "ILI9342c: Error length of Display Inversion Control (INVTR) register");     \
+	BUILD_ASSERT(DT_PROP_LEN(DT_INST(n, ilitek_ili9342c), disctrl) == ILI9342C_DISCTRL_LEN,    \
+		     "ILI9342c: Error length of display function control (DISCTRL) register");    \
+	BUILD_ASSERT(DT_PROP_LEN(DT_INST(n, ilitek_ili9342c), etmod) == ILI9342C_ETMOD_LEN,   \
+		     "ILI9342c: Error length of Entry Mode set (ETMOD) register");         \
+	BUILD_ASSERT(DT_PROP_LEN(DT_INST(n, ilitek_ili9342c), pwctrl1) == ILI9342C_PWCTRL1_LEN,    \
+		     "ILI9342c: Error length of power control 1 (PWCTRL1) register");             \
+	BUILD_ASSERT(DT_PROP_LEN(DT_INST(n, ilitek_ili9342c), pwctrl2) == ILI9342C_PWCTRL2_LEN,    \
+		     "ILI9342c: Error length of power control 2 (PWCTRL2) register");             \
+	BUILD_ASSERT(DT_PROP_LEN(DT_INST(n, ilitek_ili9342c), pwctrl3) == ILI9342C_PWCTRL3_LEN,    \
+		     "ILI9342c: Error length of power control 3 (PWCTRL3) register");             \
+	BUILD_ASSERT(DT_PROP_LEN(DT_INST(n, ilitek_ili9342c), vmctrl1) == ILI9342C_VMCTRL1_LEN,    \
+		     "ILI9342c: Error length of VCOM control 1 (VMCTRL1) register");              \
+	BUILD_ASSERT(DT_PROP_LEN(DT_INST(n, ilitek_ili9342c), pgamctrl) == ILI9342C_PGAMCTRL_LEN,  \
+		     "ILI9342c: Error length of positive gamma correction (PGAMCTRL) register");  \
+	BUILD_ASSERT(DT_PROP_LEN(DT_INST(n, ilitek_ili9342c), ngamctrl) == ILI9342C_NGAMCTRL_LEN,  \
+		     "ILI9342c: Error length of negative gamma correction (NGAMCTRL) register");  \
+	BUILD_ASSERT(DT_PROP_LEN(DT_INST(n, ilitek_ili9342c), ifctl) == ILI9342C_IFCTL_LEN,   \
+		     "ILI9342c: Error length of frame rate control (IFCTL) register");            \
```
None of this makes any sense.

ExtremeGTX replied:
it is like what is already here: https://github.com/zephyrproject-rtos/zephyr/blame/main/drivers/display/display_ili9341.h

jfischer-no replied:
Still does no convince me to litter the source code with BUILD_ASSERT and does not help you.

ExtremeGTX replied:
Done

jfischer-no commented on drivers/display/display_ili9342c.c:
```diff
@@ -0,0 +1,107 @@
+
+	/*  some commands require that SETEXTC be set first before it becomes enabled. */
+	LOG_HEXDUMP_DBG(regs->setextc, ILI9342C_SETEXTC_LEN, "SETEXTC");
+	r = ili9xxx_transmit(dev, ILI9342C_SETEXTC, regs->setextc, ILI9342C_SETEXTC_LEN);
```
```c

        r = ili9xxx_transmit(dev, ILI9342C_SETEXTC, regs->setextc, 
                             ILI9342C_SETEXTC_LEN);
```
like original drivers/display/display_ili9340.c

ExtremeGTX replied:
Done

jfischer-no commented on drivers/display/display_ili9342c.h:
```diff
@@ -0,0 +1,121 @@
+	BUILD_ASSERT(DT_PROP_LEN(DT_INST(n, ilitek_ili9342c), ifctl) == ILI9342C_IFCTL_LEN,   \
+		     "ILI9342c: Error length of frame rate control (IFCTL) register");            \
+	static const struct ili9342c_regs ili9xxx_regs_##n = {		\
+	.gamset = DT_PROP(DT_INST(n, ilitek_ili9342c), gamset),		\
+	.ifmode = DT_PROP(DT_INST(n, ilitek_ili9342c), ifmode),		\
+	.frmctr1 = DT_PROP(DT_INST(n, ilitek_ili9342c), frmctr1),	\
+	.invtr = DT_PROP(DT_INST(n, ilitek_ili9342c), invtr),		\
+	.disctrl = DT_PROP(DT_INST(n, ilitek_ili9342c), disctrl),	\
+	.etmod = DT_PROP(DT_INST(n, ilitek_ili9342c), etmod),	\
+	.pwctrl1 = DT_PROP(DT_INST(n, ilitek_ili9342c), pwctrl1),	\
+	.pwctrl2 = DT_PROP(DT_INST(n, ilitek_ili9342c), pwctrl2),	\
+	.pwctrl3 = DT_PROP(DT_INST(n, ilitek_ili9342c), pwctrl3),	\
+	.vmctrl1 = DT_PROP(DT_INST(n, ilitek_ili9342c), vmctrl1),	\
+	.setextc = {0xFF, 0x93, 0x42},								\
+	.pgamctrl = DT_PROP(DT_INST(n, ilitek_ili9342c), pgamctrl),	\
+	.ngamctrl = DT_PROP(DT_INST(n, ilitek_ili9342c), ngamctrl),	\
+	.ifctl = DT_PROP(DT_INST(n, ilitek_ili9342c), ifctl),		\
```
That does not look properly indented.

ExtremeGTX replied:
Done

jfischer-no reviewed (dismissed):
LGTM

diegoherranz commented:
Sorry I didn't have time to look into this before.
I had a look today and it looks good to me.
My only question is, what's the final state of the rotation issue?

For me, using a display with this controller, I need `ILI9XXX_MADCTL_MX | ILI9XXX_MADCTL_MY` for the orientation to be correct, and I think I wouldn't be able to do that with this driver as is (none of the 4 orientations would work). Would it work for you, @ExtremeGTX ?

Thanks!

ExtremeGTX commented:
> Sorry I didn't have time to look into this before.
> I had a look today and it looks good to me.
> My only question is, what's the final state of the rotation issue?
> 
> For me, using a display with this controller, I need `ILI9XXX_MADCTL_MX | ILI9XXX_MADCTL_MY` for the orientation to be correct, and I think I wouldn't be able to do that with this driver as is (none of the 4 orientations would work). Would it work for you, @ExtremeGTX ?
> 
> Thanks!

@diegoherranz
yes Orientation is broken, and i think the right way here to go ahead is to implement a driver quirk as @jfischer-no recommended as a workaround, similar to what is happening in ssdxxx display drivers.

He approved the PR but i'm not ok with merging without a warning or a fix.

I will try to implement the driver quirk solution in a separate branch. if the solution is fine i will push it here.

diegoherranz commented:
Thanks for the update. Let me know if you want me to test anything. I have a setup where I can test it quickly if you want to double check anything.

ExtremeGTX commented:
> Thanks for the update. Let me know if you want me to test anything. I have a setup where I can test it quickly if you want to double check anything.

@diegoherranz 
Thanks a lot, if you can test this for me please on another controller than 9342c.
When you change the orientation from 0 or 180 to 90 or 270, do you switch the display properties: `width`, `height` in the device tree ?
Example:
```cpp
		rotation = <0>;
		width = <320>;
		height = <240>;
```
vs 
```cpp
		rotation = <90>;
		width = <240>;    /* Notice the value here was the height in the previous config */
		height = <320>;  /* Notice the value here was the width in the previous config */
```

diegoherranz commented:
> if you can test this for me please on another controller than 9342c.

Unfortunately, I only have displays with the 9342c. So, probably I can only double check what you test.

ExtremeGTX commented:
@jfischer-no @diegoherranz I have added orientation support via quirks here in a separate commit, WDYT ?

diegoherranz commented:
I've just tested it and it works for me on a 9342c, (I've tested the 0 and 180 degree orientations).
Comments:
* Quite a few warnings now.
* I'm not familiar with how quirks should be implemented so I can't provide useful feedback at the moment.

Thanks!

jfischer-no commented on drivers/display/display_ili9xxx.h:
```diff
@@ -54,7 +54,18 @@
 
+enum madctl_cmd_set {
+	CMD_SET_1,	/* Default for most of ili9xxx display controllers */
+	CMD_SET_2,	/* Used by ili9342c */
```
```
	CMD_SET_1,	/* Default for most of ILI9xxx display controllers */
	CMD_SET_2,	/* Used by ILI9342c */
```

jfischer-no commented on drivers/display/display_ili9xxx.h:
```diff
@@ -54,7 +54,18 @@
+};
+
+struct ili9xxx_quirks {
+	enum madctl_cmd_set orientation_cmd_set;
```
enum madctl_cmd_set does not seem to be orientation specific, so why not just `enum madctl_cmd_set cmd_set;` ?

ExtremeGTX commented:
@galak @jfischer-no since there is always a time difference between me and you, could you please Approve and merge?
When I force push to rebase on the latest main, Github dismisses the review.

ExtremeGTX commented:
@carlescufi @jfischer-no @MaureenHelm what is the way forward ?

diegoherranz commented:
Just one comment. I tested this again today after the rebase and there are fewer warnings than last time I tested it, but it's still complaining about the unused quirks (those for the display controller models not in use). In my case, I'm using ili9342c, so it complains about the others:

```
zephyr/drivers/display/display_ili9xxx.c:435:68: warning: 'ili9488_quirks' defined but not used [-Wunused-const-variable=]
  435 | static const struct ili9xxx_quirks ili9340_quirks, ili9341_quirks, ili9488_quirks = {
      |                                                                    ^~~~~~~~~~~~~~
zephyr/drivers/display/display_ili9xxx.c:435:52: warning: 'ili9341_quirks' defined but not used [-Wunused-const-variable=]
  435 | static const struct ili9xxx_quirks ili9340_quirks, ili9341_quirks, ili9488_quirks = {
      |                                                    ^~~~~~~~~~~~~~
zephyr/drivers/display/display_ili9xxx.c:435:36: warning: 'ili9340_quirks' defined but not used [-Wunused-const-variable=]
  435 | static const struct ili9xxx_quirks ili9340_quirks, ili9341_quirks, ili9488_quirks = {
```

The code in `display_ili9xxx.c` is:

```
static const struct ili9xxx_quirks ili9340_quirks, ili9341_quirks, ili9488_quirks = {
	.cmd_set = CMD_SET_1,
};

static const struct ili9xxx_quirks ili9342c_quirks = {
	.cmd_set = CMD_SET_2,
};
```

Should we do something like?

```
#ifdef CONFIG_ILI9340
static const struct ili9xxx_quirks ili9340_quirks = {
	.cmd_set = CMD_SET_1,
};
#endif

#ifdef CONFIG_ILI9341
static const struct ili9xxx_quirks ili9341_quirks = {
	.cmd_set = CMD_SET_1,
};
#endif

#ifdef CONFIG_ILI9342C
static const struct ili9xxx_quirks ili9342c_quirks = {
	.cmd_set = CMD_SET_2,
};
#endif

#ifdef CONFIG_ILI9488
static const struct ili9xxx_quirks ili9488_quirks = {
	.cmd_set = CMD_SET_1,
};
#endif
```

Or even create one single structure and assign the `.cmd_set` using the defines? Probably finding a better name for `ili9xxx_quirks_in_use`:

```
static const struct ili9xxx_quirks ili9xxx_quirks_in_use = {
        #if defined(CONFIG_ILI9340) || defined(CONFIG_ILI9341) || defined(CONFIG_ILI9488)
 	.cmd_set = CMD_SET_1,
        #elif defined(CONFIG_ILI9342C)
        .cmd_set = CMD_SET_2,
        #endif
};

[...]
									       \
	static const struct ili9xxx_config ili9xxx_config_##n = {              \
		.quirks = &ili9xxx_quirks_in_use,  
[...]
```

Really keen to see this merged :)

Thanks!

MaureenHelm commented on drivers/display/display_ili9342c.h:
```diff
@@ -0,0 +1,95 @@
+/* Initializer macro for ILI9342C registers. */
+#define ILI9342c_REGS_INIT(n)                                           \
+	static const struct ili9342c_regs ili9xxx_regs_##n = {		\
+	.gamset = DT_PROP(DT_INST(n, ilitek_ili9342c), gamset),		\
+	.ifmode = DT_PROP(DT_INST(n, ilitek_ili9342c), ifmode),		\
+	.frmctr1 = DT_PROP(DT_INST(n, ilitek_ili9342c), frmctr1),	\
+	.invtr = DT_PROP(DT_INST(n, ilitek_ili9342c), invtr),		\
+	.disctrl = DT_PROP(DT_INST(n, ilitek_ili9342c), disctrl),	\
+	.etmod = DT_PROP(DT_INST(n, ilitek_ili9342c), etmod),		\
+	.pwctrl1 = DT_PROP(DT_INST(n, ilitek_ili9342c), pwctrl1),	\
+	.pwctrl2 = DT_PROP(DT_INST(n, ilitek_ili9342c), pwctrl2),	\
+	.pwctrl3 = DT_PROP(DT_INST(n, ilitek_ili9342c), pwctrl3),	\
+	.vmctrl1 = DT_PROP(DT_INST(n, ilitek_ili9342c), vmctrl1),	\
+	.setextc = {0xFF, 0x93, 0x42},                                  \
+	.pgamctrl = DT_PROP(DT_INST(n, ilitek_ili9342c), pgamctrl),	\
+	.ngamctrl = DT_PROP(DT_INST(n, ilitek_ili9342c), ngamctrl),	\
+	.ifctl = DT_PROP(DT_INST(n, ilitek_ili9342c), ifctl),		\
```
you can simplify with `DT_INST_PROP`

ExtremeGTX replied:
ili9xxx drivers doesn't include `DT_DRV_COMPAT` value. so I chose to follow the other variants.

MaureenHelm commented on dts/bindings/display/ilitek,ili9342c.yaml:
```diff
@@ -0,0 +1,124 @@
+properties:
+  gamset:
+    type: uint8-array
+    default: [0x01]
```
Defaults need justification in the description

MaureenHelm reviewed (dismissed):
> Just one comment. I tested this again today after the rebase and there are fewer warnings than last time I tested it, but it's still complaining about the unused quirks (those for the display controller models not in use)

Build warnings need to be addressed. Is there anything in-tree (i.e., sample or test) that builds this driver?

ExtremeGTX committed:
drivers: display: ili9342c display driver

This driver implement basic functions of ili9342c controller
which comes mostly with IPS displays.

Signed-off-by: Mohamed ElShahawi <ExtremeGTX@hotmail.com>

ExtremeGTX commented:
@diegoherranz 
I went with the first way so if a system has 2 displays with different controllers will be able to enable both driver instances.
The problem with the 2nd solution is when CONFIG_ILI9340 and CONFIG_ILI9342c are defined at the same time.

@MaureenHelm asking for including this driver in (sample/test) or maybe a board like other ili9xxx drivers.
So, Do you @diegoherranz have a devkit where we can add so it compiles this driver in CI ? right now there isn't any config in Zephyr tree which enables this driver.

diegoherranz commented:
> I went with the first way so if a system has 2 displays with different controllers will be able to enable both driver instances.
The problem with the 2nd solution is when CONFIG_ILI9340 and CONFIG_ILI9342c are defined at the same time.

Cool. It builds without warning to me now (and it works), using a ili9342c screen.


> Do you @diegoherranz have a devkit where we can add so it compiles this driver in CI ? right now there isn't any config in Zephyr tree which enables this driver

I do but it is a proprietary/personal board. I don't mind sharing its board folder (DTS, defconfig, etc.), but not sure if that's OK to include in zephyr, not being a publicly available board. Just let me know.

Thanks!

ExtremeGTX committed:
tests: drivers: build_all: Add test for display drivers

Add a blank test to cover building drivers that are in drivers/display/

Signed-off-by: Mohamed ElShahawi <ExtremeGTX@hotmail.com>
